### PR TITLE
Add Rancher app for install

### PR DESCRIPTION
**Pull request by AminArria**
This adds the rancher app for install. For now it supports installing either generating TLS certs self-signed by rancher (`--ingress-source rancher`) or by using Let's Encrypt (`--ingress-source letsEncrypt` default value). 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Actually was raised by #79 and then gave me permission to take it over in #81 

## How Has This Been Tested?
Deployed to a new cluster doing:

```
./arkade install rancher --hostname arkade.gq --letsencrypt-email arria.amin@gmail.com --replicas 1
```

```
./arkade install rancher --hostname <domain> --ingress-source rancher --replicas 1
```

And later could access the rancher interface at `<domain>`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

- [x] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

**Comment by alexellis** [comment]
@saiyam1814 please could you take a look and compare to #81? The main blocker on #81 was not using helm.

@Waterdrips can you take a look at the approach and give your take?

**Comment by AminArria** [comment]
Checking if I understand, you mean to use [this](https://github.com/alexellis/arkade/blob/5c46d647d5312c82adec01364bea07e6a6b45e8e/cmd/apps/loki_app.go#L72-L77)?

```go
		lokiOptions := types.DefaultInstallOptions().
			WithNamespace(namespace).
			WithHelmPath(path.Join(userPath, ".helm")).
			WithHelmRepo("loki/loki-stack").
			WithHelmURL("https://grafana.github.io/loki/charts").
			WithOverrides(overrides)
```

**Comment by saiyam1814** [comment]
> @saiyam1814 please could you take a look and compare to #81? The main blocker on #81 was not using helm.
> 
> @Waterdrips can you take a look at the approach and give your take?

Hostname was the issue earlier for which we were blocked so unless we give a hostname Rancher won't get deployed. I was thinking of a way to deploy it even for development purposes. But now the hostname becomes mandatory else it returns an error. This might be like if you have hostname then you can try the rancher app else not. 
Also Cert-manager check is missing, as per Rancher Docker Cer-manager has to be installed first. So arkade install cert-manager if not already there should be included.

**Comment by AminArria** [comment]
I've added the last `--ingress-source` option `secret` for using your own certificates and a couple flags to create the secrets as [explained by Rancher](https://rancher.com/docs/rancher/v2.x/en/installation/options/tls-secrets/)

@Waterdrips made the change to use chart_app let me know if its what you expected

@saiyam1814 regarding the domain, it is required because Rancher requires it. But you can put anything and then set it in your `/etc/hosts/`, no need to actually own a domain.

One thing I'm not sure is how to check for `cert-manager` before installing and install it if not present.

PS: I forgot to sign-off a middle commit, should I do something about it?

**Comment by Waterdrips** [comment]
> PS: I forgot to sign-off a middle commit, should I do something about it?

If you could squash the 3 commits into 1 signed commit that would be preferable.

something like 
` git rebase -i HEAD~3` and follow the instructions in the interactive rebase

File: cmd/apps/rancher_app.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+		if tlsCa != "" {
+			overrides["privateCA"] = "true"
+		}
+
+		// install the chart
+		helmHome := path.Join(userPath, ".helm")
+		updateRepo, _ := rancher.Flags().GetBool("update-repo")
+
+		rancherOptions := types.DefaultInstallOptions().
```
Comment (alexellis) [inline cmd/apps/rancher_app.go:145]: @Waterdrips is this more like what you were looking for?

File: cmd/apps/rancher_app.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		fmt.Println(rancherInstallMsg)
+
+		return nil
+	}
+
+	return rancher
+}
+
+const RancherInfoMsg = `
+# Wait for Rancher to be rolled out:
```
Comment (alexellis) [inline cmd/apps/rancher_app.go:179]: Please could you also add a URL for the docs / getting started page?

File: cmd/apps/rancher_app.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		Example:      "arkade install rancher --hostname mycluster.example.com ",
+		SilenceUsage: true,
+	}
+
+	rancher.Flags().StringP("namespace", "n", "cattle-system", "The namespace to install rancher")
+	rancher.Flags().Bool("update-repo", true, "Update the helm repo")
+	rancher.Flags().String("hostname", "", "public DNS record to access rancher")
+	rancher.Flags().String("ingress-source", "letsEncrypt", "source of TLS cert: letsEncrypt, rancher, secret")
+	rancher.Flags().String("letsencrypt-email", "", "email address used for communication about your certificate")
+	rancher.Flags().Int("replicas", 3, "Number of replicas of rancher pods")
```
Comment (alexellis) [inline cmd/apps/rancher_app.go:37]: Is this a production configuration, or min config? Would `1` also be usable?

File: cmd/apps/rancher_app.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	rancher.Flags().String("tls-cert", "", "Path to PEM encoded public key certificate")
+	rancher.Flags().String("tls-key", "", "Path to private key associated with given certificate")
+	rancher.Flags().String("tls-ca", "", "Path to private CA certificate")
+
+	rancher.RunE = func(command *cobra.Command, args []string) error {
+		// Get all flags
+		namespace, _ := command.Flags().GetString("namespace")
+		hostname, _ := command.Flags().GetString("hostname")
+		ingressSource, _ := command.Flags().GetString("ingress-source")
+		letsencryptEmail, _ := command.Flags().GetString("letsencrypt-email")
```
Comment (alexellis) [inline cmd/apps/rancher_app.go:47]: Is this required for Rancher to work? How do people run it locally? (inlets-operator could help here if LE is required)

**Comment by alexellis** [comment]
How popular is --ingress-source for a self-signed cert vs. LE?

File: cmd/apps/rancher_app.go (hunk 3 above)
Comment (AminArria) [reply cmd/apps/rancher_app.go:37]: `3` is the default for the helm chart and their usual recommendation for HA, but yeah you can use `1`

All my tests where with `--replicas 1`

File: cmd/apps/rancher_app.go (hunk 4 above)
Comment (AminArria) [reply cmd/apps/rancher_app.go:47]: `letsencrypt-email` is only required if using LE for ingress, but I didn't check if it actually is a hard requirement on the chart itself.

Locally you could used the `rancher` or `secret` alternatives for `ingress-source`

File: cmd/apps/rancher_app.go (hunk 2 above)
Comment (AminArria) [reply cmd/apps/rancher_app.go:182]: Added the link to docs

**Comment by AminArria** [comment]
No idea, on which is used more. Personally I've always used LE

**Comment by saiyam1814** [comment]
> I've added the last `--ingress-source` option `secret` for using your own certificates and a couple flags to create the secrets as [explained by Rancher](https://rancher.com/docs/rancher/v2.x/en/installation/options/tls-secrets/)
> 
> I tested it by creating my own certificates and CA following [this guide](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/)
> 
> @Waterdrips made the change to use chart_app let me know if its what you expected
> 
> @saiyam1814 regarding the domain, it is required because Rancher requires it. But you can put anything and then set it in your `/etc/hosts/`, no need to actually own a domain.
> 
> One thing I'm not sure is how to check for `cert-manager` before installing and install it if not present.
> 
> PS: I forgot to sign-off a middle commit, should I do something about it?

I think the best way to check for cert-manager would be to check for arkade install cert-manager (if it's done then proceed)

**Comment by alexellis** [comment]
Let's get the first release out without the check

> One thing I'm not sure is how to check for cert-manager before installing and install it if not present.

We can add a mention in the app's description that cert-manager is required, and think about this more after.

Your commits will need rebasing onto master, so please could you do that? Make sure you don't introduce a merge commit at the same time. I tend to use `git pull origin master --rebase` myself.

**Comment by alexellis** [comment]
@Waterdrips any further comments at this time?

**Comment by alexellis** [comment]
Thank you @AminArria 

Could you rebase this PR again please?

**Comment by AminArria** [comment]
@alexellis  rebase done :D

**Comment by Shikachuu** [comment]
@AminArria are you still intrested?
May I ask you to rebase the PR again?

**Comment by AminArria** [comment]
Hi @Shikachuu I'll do the rebase later this weekend

**Comment by Shikachuu** [comment]
Hey @AminArria are you still intrested to do that rebase?

**Comment by alexellis** [comment]
@AminArria are you still interested in completing this app or do you want to close your PR?

**Comment by AminArria** [comment]
Hey @alexellis yeah sorry don't have the time for it, I'll close it

**Pull request closed without merging**